### PR TITLE
XW-260

### PR DESCRIPTION
**Pull request by tomasznapieralski**
https://wikia-inc.atlassian.net/browse/XW-260

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/VerbatimTests.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.testng.annotations.Test;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @ownership: Content X-Wing
+ */
+// Uncoment after finish all verbatim tags
+//@Test(groups = {"MercuryVerbatimTests", "Mercury"})
+public class VerbatimTests extends NewTestTemplate {
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/VerbatimTests.java]: verbatim is just a name of tag in mediawiki. The tested functionality is a widget. 
change the class name to  widgetTests or thirdparties

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/VerbatimTests.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package com.wikia.webdriver.testcases.mercurytests;
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/VerbatimTests.java]: move it to package like widgets or thirdparties. Verbatim is mediawiki syntax relatex term, not functionality

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/VerbatimPageObject.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  protected abstract boolean isTagLoadedOnOasis();
+
+  protected String getTagName() {
+    return tagName;
+  }
+
+  public boolean isLoadedOnMercury() {
+    boolean result = isTagLoadedOnMercury();
+    Assertion.assertTrue(result, MercuryMessages.INVISIBLE_MSG);
+    PageObjectLogging.log(getTagName(), MercuryMessages.VISIBLE_MSG, result);
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/VerbatimPageObject.java]: :+1:  for moving logging into method, rather then adding it after method usage

File: src/test/java/com/wikia/webdriver/testcases/verbatimtests/VerbatimTests.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import com.wikia.webdriver.common.core.api.ArticleContent;
+import com.wikia.webdriver.common.templates.NewTestTemplate;
+import com.wikia.webdriver.pageobjectsfactory.pageobject.verbatim.widgets.TwitterPageObject;
+
+import org.testng.annotations.Test;
+
+/**
+ * @ownership: Content X-Wing
+ */
+public class VerbatimTests extends NewTestTemplate {
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/testcases/verbatimtests/VerbatimTests.java]: verbatim is just a name of tag in mediawiki. The tested functionality is a widget. 
change the class name to widgetTests or thirdparties

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/widgets/TwitterPageObject.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package com.wikia.webdriver.pageobjectsfactory.pageobject.verbatim.widgets;
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/widgets/TwitterPageObject.java]: Classifying page objects by functionality  not by syntax terms. Change the verbatim part of the package name to thirdparties or widgets and limit the number of levels from 2 to 1. There is no need to make two subfolders

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/VerbatimPageObject.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package com.wikia.webdriver.pageobjectsfactory.pageobject.verbatim;
```
Comment (justnpT) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/verbatim/VerbatimPageObject.java]: Classifying page objects by functionality  not by syntax terms is the way to go. Change the verbatim part of the package name to thirdparties or widgets

**Comment by justnpT** [comment]
lgtm

File: src/test/java/com/wikia/webdriver/common/contentpatterns/MercurySubpages.java [hunk 7]
```diff
@@ -17,6 +17,8 @@
   public static final String QUESTION_MARK = "Question?mark?question";
   public static final String COLON = "Colon:colon:colon";
 
+  public static final String TWITTER_WIDGET = "TwitterWidget";
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/common/contentpatterns/MercurySubpages.java]: do we need this?

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  protected abstract boolean isTagLoadedOnOasis();
+
+  protected String getTagName() {
+    return tagName;
+  }
+
+  public boolean isLoadedOnMercury() {
+    boolean result = isTagLoadedOnMercury();
+    Assertion.assertTrue(result, MercuryMessages.INVISIBLE_MSG);
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java]: assertion should made in test. Page objects are from keeping page state not asserting if something is correct or not

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/WidgetTests.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    // This login is temporary solution, use @Execute after QAART-669 is done
+    new LoginPage(driver).get().logUserIn(Configuration.getCredentials().userNameStaff2,
+                                          Configuration.getCredentials().passwordStaff2);
+  }
+
+  @Test(groups = "MercuryWidgetTest_001")
+  @Execute(onWikia = "mercuryautomationtesting")
+  public void MercuryWidgetTest_001_Twitter() {
+    ArticleContent articleContent = new ArticleContent();
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/WidgetTests.java]: consider moving those functions to TwitterWidgetPageObject and name the method:
```
createPageWithWidget
```

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define and throw a dedicated exception instead of using a generic one.](https://github.com/Wikia/selenium-tests/blob/ca3fe9ad990fd671033440cfecc3b2dcba6eea85/src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java#L68) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)
* MAJOR [Rename this class name to match the regular expression '^[A-Z][a-zA-Z0-9]*$'.](https://github.com/Wikia/selenium-tests/blob/ca3fe9ad990fd671033440cfecc3b2dcba6eea85/src/test/java/com/wikia/webdriver/testcases/createawikitests/CreateWikiTests_lang.java#L20) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00101)
* MAJOR [This block of commented-out lines of code should be removed.](https://github.com/Wikia/selenium-tests/blob/ca3fe9ad990fd671033440cfecc3b2dcba6eea85/src/test/java/com/wikia/webdriver/testcases/mediatests/lightboxtests/LightboxTests.java#L177) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* MINOR [Move this trailing comment on the previous empty line.](https://github.com/Wikia/selenium-tests/blob/ca3fe9ad990fd671033440cfecc3b2dcba6eea85/src/test/java/com/wikia/webdriver/testcases/mediatests/lightboxtests/LightboxTests.java#L104) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ATrailingCommentCheck)
* INFO [Complete the task associated to this TODO comment.](https://github.com/Wikia/selenium-tests/blob/ca3fe9ad990fd671033440cfecc3b2dcba6eea85/src/test/java/com/wikia/webdriver/testcases/mercurytests/curatedcontenttests/CropImageTests.java#L38) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java [hunk 10]
```diff
@@ -3,12 +3,10 @@
...
 import com.wikia.webdriver.common.core.CommonUtils;
 import com.wikia.webdriver.common.logging.PageObjectLogging;
+
 import org.apache.commons.lang.exception.ExceptionUtils;
 
-import java.awt.Color;
-import java.awt.Font;
-import java.awt.FontMetrics;
-import java.awt.Graphics2D;
+import java.awt.*;
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java]: don't use * in imports

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  protected abstract boolean isTagLoadedOnMercury();
+
+  protected abstract boolean isTagLoadedOnOasis();
+
+  protected String getTagName() {
+    return tagName;
+  }
+
+  private String getArticleName() {
+    return tagName.substring(0, 1).toUpperCase() + tagName.substring(1).toLowerCase() + "Widget";
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java]: article name and tag (<tag>) could be defined as fields in TwitterWidget class so it's more verbose

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  private String getTag() {
+    return "<" + getTagName() + ">";
+  }
+
+  public void createAndNavigate(String wikiUrl) {
+    ArticleContent articleContent = new ArticleContent();
+    articleContent.clear(getArticleName());
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/WidgetPageObject.java]: you can assign return of getArticleName to variable

**Comment by bognix** [comment]
resolve conflicts before another round of cr

File: src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java (hunk 10 above)
Comment (tomasznapieralski) [inline src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java]: true, i've used 'optimalize imports' option in Intellij
i found that wildcard option can be disabled

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define and throw a dedicated exception instead of using a generic one.](https://github.com/Wikia/selenium-tests/blob/d8dbb6dcd744cc003253c6935fa5b54eb033457c/src/test/java/com/wikia/webdriver/common/core/imageutilities/ImageGenerator.java#L71) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)
* MAJOR [Rename this class name to match the regular expression '^[A-Z][a-zA-Z0-9]*$'.](https://github.com/Wikia/selenium-tests/blob/d8dbb6dcd744cc003253c6935fa5b54eb033457c/src/test/java/com/wikia/webdriver/testcases/createawikitests/CreateWikiTests_lang.java#L20) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00101)
* MINOR [Move this trailing comment on the previous empty line.](https://github.com/Wikia/selenium-tests/blob/d8dbb6dcd744cc003253c6935fa5b54eb033457c/src/test/java/com/wikia/webdriver/testcases/mediatests/lightboxtests/LightboxTests.java#L104) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ATrailingCommentCheck)
* MAJOR [This block of commented-out lines of code should be removed.](https://github.com/Wikia/selenium-tests/blob/d8dbb6dcd744cc003253c6935fa5b54eb033457c/src/test/java/com/wikia/webdriver/testcases/mediatests/lightboxtests/LightboxTests.java#L177) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* INFO [Complete the task associated to this TODO comment.](https://github.com/Wikia/selenium-tests/blob/d8dbb6dcd744cc003253c6935fa5b54eb033457c/src/test/java/com/wikia/webdriver/testcases/mercurytests/curatedcontenttests/CropImageTests.java#L38) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    return TAG_NAME;
+  }
+
+  protected String getTag() {
+    return "<" + getTagName() + ">";
+  }
+
+  protected boolean isTagLoadedOnMercury() {
+    // to be implemented
+    return isElementVisible(elementInVebatim);
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java]: how about using https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/NotImplementedException.html?

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  protected boolean isTagLoadedOnOasis() {
+    // to be implemented
+    return isElementVisible(elementInVebatim);
```
Comment (bognix) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java]: how about using https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/NotImplementedException.html

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    return TAG_NAME;
+  }
+
+  protected String getTag() {
+    return "<" + getTagName() + ">";
+  }
+
+  protected boolean isTagLoadedOnMercury() {
+    throw new NotImplementedException();
+    //return isElementVisible(elementInVebatim);
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
+package com.wikia.webdriver.pageobjectsfactory.pageobject.widget;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java]: MAJOR Replace this usage of Sun classes by ones from the Java API. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1191)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  protected boolean isTagLoadedOnMercury() {
+    throw new NotImplementedException();
+    //return isElementVisible(elementInVebatim);
+  }
+
+  protected boolean isTagLoadedOnOasis() {
+    throw new NotImplementedException();
+    //return isElementVisible(elementInVebatim);
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/widget/TwitterWidgetPageObject.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.

**Pull request merged**